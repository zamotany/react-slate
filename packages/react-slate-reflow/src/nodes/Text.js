/* @flow */

import assert from 'assert';
import UnitLayout from '../layout/elements/UnitLayout';
import type { Node } from '../types';

export default class Text implements Node {
  body: string = '';
  parent = null;
  // Children should always be empty.
  children = Object.freeze([]);
  layout = new UnitLayout(this);

  setBody(body: string) {
    assert(!body.includes('\n'), 'Text body cannot have new line characters');
    this.body = body;
  }
}
