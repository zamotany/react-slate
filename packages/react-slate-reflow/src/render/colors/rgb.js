/* @flow */

import memoize from 'fast-memoize';
import supportsColor from 'supports-color';
import convert from 'color-convert';

const CSI = '\u001b[';
const RGB_OPEN = `${CSI}38`;
const RGB_BG_OPEN = `${CSI}48`;
const COLOR_CLOSE = `${CSI}39m`;
const BG_COLOR_CLOSE = `${CSI}49m`;
const BG_COLOR_OFFSET = 10;

export function create(level: number) {
  return {
    withRgbColor: memoize(
      {
        '1': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) => {
          const code = convert.rgb.ansi16(rgb);
          return `${CSI}${code}m${text}${COLOR_CLOSE}`;
        },
        '2': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) => `${RGB_OPEN};5;${convert.rgb.ansi256(rgb)}m${text}${COLOR_CLOSE}`,
        '3': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) =>
          `${RGB_OPEN};2;${rgb[0]};${rgb[1]};${rgb[2]}m${text}${COLOR_CLOSE}`,
      }[level]
    ),
    withRgbBackgroundColor: memoize(
      {
        '1': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) => {
          const code = convert.rgb.ansi16(rgb);
          return `${CSI}${code + BG_COLOR_OFFSET}m${text}${BG_COLOR_CLOSE}`;
        },
        '2': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) =>
          `${RGB_BG_OPEN};5;${convert.rgb.ansi256(
            rgb
          )}m${text}${BG_COLOR_CLOSE}`,
        '3': (
          rgb: [number | string, number | string, number | string],
          text: string
        ) =>
          `${RGB_BG_OPEN};2;${rgb[0]};${rgb[1]};${
            rgb[2]
          }m${text}${BG_COLOR_CLOSE}`,
      }[level]
    ),
  };
}

export const { withRgbColor, withRgbBackgroundColor } = create(
  process.env.CI ? 1 : supportsColor.stdout.level
);
