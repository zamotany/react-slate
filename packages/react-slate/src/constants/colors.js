/* @flow */

export default {
  black: 'ansi-black',
  red: 'ansi-red',
  green: 'ansi-green',
  yellow: 'ansi-yellow',
  blue: 'ansi-blue',
  magenta: 'ansi-magenta',
  cyan: 'ansi-cyan',
  white: 'ansi-white',
  gray: 'ansi-gray',
};
