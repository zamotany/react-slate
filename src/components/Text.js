// @flow

import React, { Children } from 'react';
import chalk from 'chalk';
import { Chunk, Endl } from './';

type Style = {|
  color?: string,
  backgroundColor?: string,
  fontWeight?: 'bold' | 'normal',
  fontStyle?: 'italic' | 'normal',
  textDecoration?: 'underline' | 'line-through' | 'normal',
  textTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase',
  visibility?: 'visible' | 'hidden',
|};

type Props = {
  style?: Style,
  endl?: boolean,
  children: any,
};

function capitalize(text: string) {
  return text
    .split(' ')
    .map(t => t.charAt(0).toUpperCase() + t.slice(1))
    .join(' ');
}

function stylize(style: Style, text: string) {
  /* eslint-disable no-param-reassign */

  let color = style.color;
  let inverse = false;

  if (style.backgroundColor) {
    color = style.backgroundColor;
    inverse = true;
  }

  let enhancer = chalk.reset;

  if (color) {
    if (color.startsWith('#')) {
      enhancer = enhancer.hex(color);
    } else if (color.startsWith('rgb')) {
      const rgb = color
        .replace(/[rgb()]/g, '')
        .split(',')
        .map(i => parseInt(i, 10));
      enhancer = chalk.rgb(...rgb);
    } else {
      enhancer = chalk.keyword(color);
    }
  }

  if (inverse) {
    enhancer = enhancer.inverse;
  }

  if (style.fontWeight === 'bold') {
    enhancer = enhancer.bold;
  }

  if (style.fontStyle === 'italic') {
    enhancer = enhancer.italic;
  }

  if (style.textDecoration) {
    // eslint-disable-next-line default-case
    switch (style.textDecoration) {
      case 'underline':
        enhancer = enhancer.underline;
        break;
      case 'line-through':
        enhancer = enhancer.strikethrough;
        break;
    }
  }

  if (style.visibility === 'hidden') {
    enhancer = enhancer.hidden;
  }

  if (style.textTransform) {
    // eslint-disable-next-line default-case
    switch (style.textTransform) {
      case 'uppercase':
        text = text.toUpperCase();
        break;
      case 'lowercase':
        text = text.toLowerCase();
        break;
      case 'capitalize':
        text = capitalize(text);
        break;
    }
  }

  return enhancer(text);
}

export default function Text(props: Props) {
  const { children, style, endl } = props;

  return (
    <Chunk>
      {style
        ? Children.map(children, child => {
            if (typeof child === 'string' || typeof child === 'number') {
              return stylize(style, String(child));
            }

            if (child && child.type === Text) {
              return React.cloneElement(child, {
                style: child.props.style
                  ? { ...style, ...child.props.style }
                  : style,
              });
            }

            return child;
          })
        : children}
      {endl ? <Endl /> : null}
    </Chunk>
  );
}
