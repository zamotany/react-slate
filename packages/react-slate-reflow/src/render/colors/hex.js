/* @flow */

import memoize from 'fast-memoize';
import convert from 'color-convert';
import supportsColor from 'supports-color';
import { create as createRgb } from './rgb';

export function create(level: number) {
  const { withRgbColor, withRgbBackgroundColor } = createRgb(level);
  return {
    withHexColor: memoize((color: string, text: string) =>
      withRgbColor(convert.hex.rgb(color), text)
    ),

    withHexBackgroundColor: memoize((color: string, text: string) =>
      withRgbBackgroundColor(convert.hex.rgb(color), text)
    ),
  };
}

export const { withHexColor, withHexBackgroundColor } = create(
  process.env.CI ? 1 : supportsColor.stdout.level
);
