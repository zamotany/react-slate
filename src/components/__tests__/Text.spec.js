/* @flow */

import * as React from 'react';
import { render } from 'enzyme';
import Text from '../Text';
import colors from '../../constants/colors';

describe('Text component', () => {
  it('renders children', () => {
    const element = render(
      <Text>
        {'Hello world\n'}
        <Text endl>Something more</Text>
        {[
          <Text key="1" endl>
            No way
          </Text>,
          <Text key="2">Haha</Text>,
        ]}
      </Text>
    );
    expect(element.text()).toMatchSnapshot();
  });

  it('appends new line', () => {
    const element = render(<Text endl>Hello world</Text>);
    expect(element.text()).toBe('Hello world\n');
  });

  it('respects color and background color', () => {
    const colored = render(
      <Text
        style={{
          color: colors.magenta,
          backgroundColor: colors.cyan,
        }}
      >
        Hello world
      </Text>
    );
    expect(colored.text()).toMatchSnapshot();
  });

  it('respects fontWeight', () => {
    const bold = render(
      <Text style={{ fontWeight: 'bold' }}>Hello world</Text>
    );
    const normal = render(
      <Text style={{ fontWeight: 'normal' }}>Hello world</Text>
    );
    expect(bold.text()).toMatchSnapshot();
    expect(normal.text()).toMatchSnapshot();
  });

  it('respects fontStyle', () => {
    const italic = render(
      <Text style={{ fontStyle: 'italic' }}>Hello world</Text>
    );
    const normal = render(
      <Text style={{ fontStyle: 'normal' }}>Hello world</Text>
    );
    expect(italic.text()).toMatchSnapshot();
    expect(normal.text()).toMatchSnapshot();
  });

  it('respects textDecoration', () => {
    const underline = render(
      <Text style={{ textDecoration: 'underline' }}>Hello world</Text>
    );
    const lineThrough = render(
      <Text style={{ textDecoration: 'line-through' }}>Hello world</Text>
    );
    const normal = render(
      <Text style={{ textDecoration: 'normal' }}>Hello world</Text>
    );
    expect(underline.text()).toMatchSnapshot();
    expect(lineThrough.text()).toMatchSnapshot();
    expect(normal.text()).toMatchSnapshot();
  });

  it('respects textTransform', () => {
    const capitalize = render(
      <Text style={{ textTransform: 'capitalize' }}>Hello world</Text>
    );
    const uppercase = render(
      <Text style={{ textTransform: 'uppercase' }}>Hello world</Text>
    );
    const lowercase = render(
      <Text style={{ textTransform: 'lowercase' }}>Hello world</Text>
    );
    const none = render(
      <Text style={{ textTransform: 'none' }}>Hello world</Text>
    );
    expect(capitalize.text()).toMatchSnapshot();
    expect(uppercase.text()).toMatchSnapshot();
    expect(lowercase.text()).toMatchSnapshot();
    expect(none.text()).toMatchSnapshot();
  });

  it('respects visibility', () => {
    const visible = render(
      <Text style={{ visibility: 'visible' }}>Hello world</Text>
    );
    const hidden = render(
      <Text style={{ visibility: 'hidden' }}>Hello world</Text>
    );
    expect(visible.text()).toMatchSnapshot();
    expect(hidden.text()).toMatchSnapshot();
  });
});
