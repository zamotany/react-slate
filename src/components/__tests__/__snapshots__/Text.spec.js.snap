// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component renders children 1`] = `
"Hello world
Something more
No way
Haha"
`;

exports[`Text component respects color and background color 1`] = `"[0m[35m[46mHello world[49m[39m[0m"`;

exports[`Text component respects fontStyle 1`] = `"[0m[3mHello world[23m[0m"`;

exports[`Text component respects fontStyle 2`] = `"[0mHello world[0m"`;

exports[`Text component respects fontWeight 1`] = `"[0m[1mHello world[22m[0m"`;

exports[`Text component respects fontWeight 2`] = `"[0mHello world[0m"`;

exports[`Text component respects textDecoration 1`] = `"[0m[4mHello world[24m[0m"`;

exports[`Text component respects textDecoration 2`] = `"[0m[9mHello world[29m[0m"`;

exports[`Text component respects textDecoration 3`] = `"[0mHello world[0m"`;

exports[`Text component respects textTransform 1`] = `"[0mHello World[0m"`;

exports[`Text component respects textTransform 2`] = `"[0mHELLO WORLD[0m"`;

exports[`Text component respects textTransform 3`] = `"[0mhello world[0m"`;

exports[`Text component respects textTransform 4`] = `"[0mHello world[0m"`;

exports[`Text component respects visibility 1`] = `"[0mHello world[0m"`;

exports[`Text component respects visibility 2`] = `"[0m[8mHello world[28m[0m"`;
